python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,255 @@
+
+        if not hasZ:
+            feedback.reportError(self.tr('The layer does not have Z values. If you have a DEM, use the Drape algorithm to extract Z values.'))
+            return
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        fieldnumber = 0
+
+        # Skip fields with names that are equal to the generated ones
+        for field in source.fields():
+            if field.name() == self.CLIMBATTRIBUTE:
+                feedback.pushWarning(self.tr(
+                                '{attr_name} attribute found and removed'.format(
+                                    attr_name=self.CLIMBATTRIBUTE)
+                                    )
+                                )
+                climbindex = fieldnumber
+            elif field.name() == self.DESCENTATTRIBUTE:
+                feedback.pushWarning(self.tr(
+                                '{attr_name} attribute found and removed'.format(
+                                    attr_name=self.DESCENTATTRIBUTE)
+                                    )
+                                )
+                descentindex = fieldnumber
+            else:
+                thefields.append(field)
+            fieldnumber = fieldnumber + 1
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        layerwithz = source
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context, thefields,
+                                               layerwithz.wkbType(),
+                                               source.sourceCrs())
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        no_z_nodes = []
+
+        for current, feature in enumerate(features):
+            # Stop the algorithm if cancelled
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            parts = feature.geometry().constParts()
+            for part in parts:
+                # Calculate the climb
+                first = True
+                zval = 0
+                for idx, v in enumerate(part.vertices()):
+                    zval = v.z()
+                    if math.isnan(zval):
+                        no_z_nodes.append(idx)
```

Comment:
I think it it too much to report individual points that miss Z values, but if that is of interest we need to do something like this (and also add the partnumber variable):

    no_z_nodes.append('Feature: ' + str(feature.id() + ' part: ' + str(partnumber) + ' point: ' + str(idx))

Reply:
Prefer a template based approach `'Feature: {feature_id}, Part: {part_id} ....'.format(feature.id(), partnumber)`, it will make translation easier (and code more readable).

Reply:
Big +1 for template, much easier for translations indeed